ooni/oonicli.py

```diff
@@ -315,3 +329,231 @@ def start():
+
+
+
+        # Wait until director has started up (including bootstrapping Tor)
+        # before adding tests
+        def post_director_start(_):
+            for net_test_loader in deck.netTestLoaders:
+                # Decks can specify different collectors
+                # for each net test, so that each NetTest
+                # may be paired with a test_helper and its collector
+                # However, a user can override this behavior by
+                # specifying a collector from the command-line (-c).
+                # If a collector is not specified in the deck, or the
+                # deck is a singleton, the default collector set in
+                # ooniprobe.conf will be used
+
+                collector = None
+                if not global_options['no-collector']:
+                    if global_options['collector']:
+                        collector = global_options['collector']
+                    elif 'collector' in config.reports \
+                            and config.reports['collector']:
+                        collector = config.reports['collector']
+                    elif net_test_loader.collector:
+                        collector = net_test_loader.collector
+
+                if collector and collector.startswith('httpo:') \
+                        and (not (config.tor_state or config.tor.socks_port)):
+                    raise errors.TorNotRunning
+
+                test_details = net_test_loader.testDetails
+                test_details['annotations'] = global_options['annotations']
+
+                director.startNetTest(net_test_loader,
+                                      global_options['reportfile'],
+                                      collector)
+            return director.allTestsDone
+
+        def start():
+            d.addCallback(setup_nettest)
+            d.addCallback(post_director_start)
+            d.addErrback(director_startup_handled_failures)
+            d.addErrback(director_startup_other_failures)
+            return d
+
+        return start()
+
+
+    finished = defer.Deferred()
+
+    @defer.inlineCallbacks
+    def readmsg(_,channel, queue_object, consumer_tag):
+        global counter, lifetime
+
+        # Wait for a message and decode it.
+        if counter >= lifetime:
+            queue_object.close(LifetimeExceeded())
+            yield channel.basic_cancel(consumer_tag=consumer_tag)
+            finished.callback(None)
+
+        else:
+            log.msg("Waiting for message")
+            
+            try:
+                ch, method, properties, body = yield queue_object.get()
+                log.msg("Got message")
+                data = json.loads(body)
+                counter += 1
+
+                log.msg("Received %d/%d: %s" %(counter, lifetime, data['url'],))
+                # acknowledge the message
+                ch.basic_ack(delivery_tag=method.delivery_tag)
+
+                d = createDeck(url=data['url'].encode('utf8'))
```

Comment:
createDeck finishes on line 488, with return start(), and returns the deferred director.allTestsDone.  We could get rid of the start() nested function entirely, I think, and put the four lines it contains into createDeck itself.